modules/javafx.graphics/src/main/java/javafx/scene/layout/BackgroundFill.java

```diff
@@ -103,6 +112,51 @@ public BackgroundFill(@NamedArg("fill") Paint fill, @NamedArg("radii") CornerRad
         hash = result;
     }
 
+    /**
+     * {@inheritDoc}
+     *
+     * @throws NullPointerException {@inheritDoc}
+     * @since 24
+     */
+    @Override
+    public BackgroundFill interpolate(BackgroundFill endValue, double t) {
+        Objects.requireNonNull(endValue, "endValue cannot be null");
+
+        // We don't check equals(endValue) here to prevent unnecessary equality checks,
+        // and only check for equality with 'this' or 'endValue' after interpolation.
+        if (t <= 0) {
+            return this;
+        }
+
+        if (t >= 1) {
+            return endValue;
+        }
+
+        // CornerRadii, Insets and Paint are implemented such that interpolate() always returns the
+        // existing instance if the intermediate value is equal to the start value or the end value,
+        // which allows us to use an identity comparison in place of a value comparison to determine
+        // equality.
+        CornerRadii newRadii = radii.interpolate(endValue.radii, t);
+        Insets newInsets = insets.interpolate(endValue.insets, t);
+        Paint newFill = PaintUtils.interpolate(fill, endValue.fill, t);
+
+        if (isSame(newFill, newRadii, newInsets)) {
+            return this;
+        }
+
+        if (endValue.isSame(newFill, newRadii, newInsets)) {
+            return endValue;
+        }
+
+        return new BackgroundFill(newFill, newRadii, newInsets);
+    }
+
+    private boolean isSame(Paint fill, CornerRadii radii, Insets insets) {
+        return this.fill == fill
+            && this.radii == radii
+            && this.insets == insets;
```

Comment:
are you sure these don't need equals()?

Reply:
This is an internal optimization, see the comment at L135. Since we control the implementation of all types here, we _know_ that these types will return the existing instance if the intermediate value is equal to either the start value or the end value. Using `equals()` here is not necessary, and is not required for correctness.